das/aggregator.go

```diff
@@ -0,0 +1,218 @@
+//
+// Copyright 2022, Offchain Labs, Inc. All rights reserved.
+//
+
+package das
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/bits"
+	"time"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/offchainlabs/nitro/arbstate"
+	"github.com/offchainlabs/nitro/blsSignatures"
+)
+
+type AggregatorConfig struct {
+	assumedHonest   int
+	retentionPeriod time.Duration
+}
+
+type Aggregator struct {
+	config   AggregatorConfig
+	services []serviceDetails
+
+	/// calculated fields
+	requiredServicesForStore       int
+	maxAllowedServiceStoreFailures int
+}
+
+type serviceDetails struct {
+	service    DataAvailabilityService
+	pubKey     blsSignatures.PublicKey
+	signerMask uint64
+}
+
+func NewAggregator(config AggregatorConfig, services []serviceDetails) *Aggregator {
+	return &Aggregator{
+		config:                         config,
+		services:                       services,
+		requiredServicesForStore:       len(services) + 1 - config.assumedHonest,
+		maxAllowedServiceStoreFailures: config.assumedHonest - 1,
+	}
+}
+
+// Retrieve calls  on each backend DAS in parallel and returns immediately on the
+// first successful response where the data matches the requested hash. Otherwise
+// if all requests fail or if its context is canceled (eg via DeadlineWrapper) then
+// it returns an error.
+func (a *Aggregator) Retrieve(ctx context.Context, cert []byte) ([]byte, error) {
+	requestedCert, _, err := arbstate.DeserializeDASCertFrom(cert)
+	if err != nil {
+		return nil, err
+	}
+
+	// Cert is the aggregate cert, validate it against DAS public keys
+	var servicesThatSignedCert []serviceDetails
+	var pubKeys []blsSignatures.PublicKey
+	for _, d := range a.services {
+		if requestedCert.SignersMask&d.signerMask != 0 {
+			servicesThatSignedCert = append(servicesThatSignedCert, d)
+			pubKeys = append(pubKeys, d.pubKey)
+		}
+	}
+	signedBlob := serializeSignableFields(*requestedCert)
+	sigMatch, err := blsSignatures.VerifySignature(requestedCert.Sig, signedBlob, blsSignatures.AggregatePublicKeys(pubKeys))
+	if err != nil {
+		return nil, err
+	}
+	if !sigMatch {
+		return nil, errors.New("Signature of data in cert passed in doesn't match")
+	}
+
+	blobChan := make(chan []byte)
```

Comment:
We probably want to make the `errorChan` buffered with size `len(servicesThatSignedCert)`, and `blobChan` buffered with size 1.

Also we probably want to `defer close(.)` these channels.

Reply:
Thanks, I think my code had a goroutine and channel leak as the first blob that was successfully returned would cause the consuming function to return, leaving the remaining routines for each backend DAS stuck trying to write to that channel.

I think that both channels need to be able to accept up to `len(servicesThatSignedCert)` items so that the goroutine for each backend DAS is able to run to completion without blocking, and then we can rely on garbage collection to remove the channels when they are no longer referenced.

I'm not sure that we actually can explicitly close them here without triggering a race between the backend querying goroutines writing to the channel and Retrieve returning.

Reply:
The cleanest way to handle this (which might be less efficient) is probably to use a waitgroup to coordinate the shutdown of the sub-calls.  Then we could have a single `defer func() { ...` to shut everything down before exiting this function. That defer would cancel the context, wait for waitgroup to complete, then close the channels.